pkg/client/client.go

```diff
@@ -632,29 +649,106 @@ func (c *Client) GetAlertingRule(ctx context.Context, namespace, name string) (*
+	}
+	unstructuredRequiredPrometheus := &unstructured.Unstructured{}
+	unstructuredRequiredPrometheus.SetUnstructuredContent(unstructuredRequiredPrometheusObject)
+
+	// Set the GVK for the unstructured object, since these are not inferred automatically, unlike their structured counterparts.
+	unstructuredRequiredPrometheus.SetGroupVersionKind(monv1.SchemeGroupVersion.WithKind(monv1.PrometheusesKind))
+
+	// Preserve the original behavior: always merge the metadata, never replace.
+	// Refer: https://github.com/openshift/cluster-monitoring-operator/pull/942.
+	prometheusGVR := unstructuredRequiredPrometheus.GroupVersionKind().GroupVersion().WithResource(monv1.PrometheusName)
+	unstructuredExistingPrometheus, err := c.dclient.
+		Resource(prometheusGVR).
+		Namespace(structuredRequiredPrometheus.GetNamespace()).
+		Get(ctx, structuredRequiredPrometheus.GetName(), metav1.GetOptions{})
 	if apierrors.IsNotFound(err) {
-		_, err := pclient.Create(ctx, p, metav1.CreateOptions{})
+		_, err = c.dclient.
+			Resource(prometheusGVR).
+			Namespace(structuredRequiredPrometheus.GetNamespace()).
+			Create(ctx, unstructuredRequiredPrometheus, metav1.CreateOptions{})
 		if err != nil {
-			return fmt.Errorf("creating Prometheus object failed: %w", err)
+			return nil, fmt.Errorf("creating Prometheus object failed: %w", err)
 		}
-		return nil
+		return ptr.To(true), nil
 	}
 	if err != nil {
-		return fmt.Errorf("retrieving Prometheus object failed: %w", err)
+		return nil, fmt.Errorf("retrieving Prometheus object failed: %w", err)
 	}
+	unstructuredRequiredPrometheusMetadataLabels := unstructuredRequiredPrometheus.GetLabels()
+	unstructuredRequiredPrometheusMetadataAnnotations := unstructuredRequiredPrometheus.GetAnnotations()
+	mergeMetadataLabels(unstructuredRequiredPrometheusMetadataLabels, unstructuredExistingPrometheus.GetLabels())
+	mergeMetadataAnnotations(unstructuredRequiredPrometheusMetadataAnnotations, unstructuredExistingPrometheus.GetAnnotations())
+	unstructuredRequiredPrometheus.SetLabels(unstructuredRequiredPrometheusMetadataLabels)
+	unstructuredRequiredPrometheus.SetAnnotations(unstructuredRequiredPrometheusMetadataAnnotations)
+	unstructuredRequiredPrometheus.SetResourceVersion(unstructuredExistingPrometheus.GetResourceVersion())
 
-	required := p.DeepCopy()
-	mergeMetadata(&required.ObjectMeta, existing.ObjectMeta)
+	_, didUpdate, err := resourceapply.ApplyUnstructuredResourceImproved(
+		ctx,
+		c.dclient,
+		c.eventRecorder,
+		unstructuredRequiredPrometheus,
+		c.resourceCache,
+		prometheusGVR,
+		prometheusDefaultingFunc,
+		nil,
+	)
+	if err != nil {
+		return &didUpdate, fmt.Errorf("updating Prometheus object failed: %w", err)
+	}
 
-	required.ResourceVersion = existing.ResourceVersion
-	_, err = pclient.Update(ctx, required, metav1.UpdateOptions{})
+	return &didUpdate, nil
+}
+
+func prometheusDefaultingFunc(unstructuredPrometheus *unstructured.Unstructured) {
```

Comment:
I thought that after our discussion here https://github.com/openshift/library-go/pull/1575#discussion_r1634427896, we "agreed" to give the "cache approach" a try for now, as it's simpler (less code/less maintenance), and with the cache, IIUC, I don't think we need to care about the defaults (if the wanted config is the one stored in the cache which mean the one deployed, skip the update).
Any reasons behind the change?

Reply:
Referring to the same thread: https://github.com/openshift/library-go/pull/1575#discussion_r1650646219, we do this in addition to the cache approach since that doesn't cache `spec` fields, in case of a metadata cache miss (same as what Trevor's patch targets in https://github.com/openshift/cluster-monitoring-operator/commit/ef5f348a939419818f9da862ff554af8853c9738#diff-2895f30116602d8fe6b0545c186d4f81247fa4eaab049dab174c6c91c0036e08R639-R667).